### PR TITLE
feat: time hooks for query experiment

### PR DESCRIPTION
**Pull request by Victor-C-Zhang**
- Merge time hooks from continuous query branch
- Ensure supernodes are restored if a continuous query errors out
- Bugfixes and improvements:
  - Reversed order of parameters in `merge_supernodes`
  - Use `const&` references for vectors
  - Improve robustness of `sizeof` calculations in binary IO operations
  - etc.

### CODE REVIEW

File: include/graph.h [hunk 1]
```diff
@@ -101,8 +101,9 @@ class Graph {
    */
   void write_binary(const std::string &filename);
 
-  std::chrono::steady_clock::time_point flush_call;
-  std::chrono::steady_clock::time_point flush_return;
+  // time hooks for experiments
+  std::chrono::steady_clock::time_point flush_start;
```
Comment (etwest) [inline include/graph.h:157]: Why did we rename these

File: src/graph.cpp [hunk 2]
```diff
@@ -217,7 +221,7 @@ std::vector<std::set<node_id_t>> Graph::boruvka_emulation() {
   for (const auto& it : temp) retval.push_back(it.second);
 
   cc_alg_end = std::chrono::steady_clock::now();
-  printf("CC done\n");
+  printf("CC done in %d rounds\n", num_rounds);
```
Comment (etwest) [inline src/graph.cpp]: Probably not necessary

File: test/simulation.cpp [hunk 3]
```diff
@@ -0,0 +1,43 @@
+#include <random>
```
Comment (etwest) [inline test/simulation.cpp]: What exactly is this simulation? A stress test of the std priority queue?

File: test/temp.cpp [hunk 4]
```diff
@@ -0,0 +1,28 @@
+#include <iostream>
```
Comment (etwest) [inline test/temp.cpp]: What is this file?

File: test/res/kron17_trouble.txt [hunk 5]
```diff
@@ -0,0 +1,9 @@
+131072  8
```
Comment (etwest) [inline test/res/kron17_trouble.txt]: What is this file for?

File: test/simulation.cpp (hunk 3 above)
Comment (Victor-C-Zhang) [reply test/simulation.cpp]: oops, an unnecessary

File: test/temp.cpp (hunk 4 above)
Comment (Victor-C-Zhang) [reply test/temp.cpp]: same

File: test/res/kron17_trouble.txt (hunk 5 above)
Comment (Victor-C-Zhang) [reply test/res/kron17_trouble.txt]: same

File: src/graph.cpp (hunk 2 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp]: I used this for some experiments, but I'm not too attached to it if you don't like it

File: include/graph.h (hunk 1 above)
Comment (Victor-C-Zhang) [reply include/graph.h:157]: for consistency

**Comment by etwest** [comment]
This looks good. I'll get Ethan's changes in for splitting cc up into multiple functions and then I'll push a commit here to ensure this is done correctly once those changes are made.

File: include/graph.h (hunk 1 above)
Comment (etwest) [reply include/graph.h:130]: Okay

Comment (etwest) [reply include/graph.h:157]: Okay

File: src/graph.cpp (hunk 2 above)
Comment (etwest) [reply src/graph.cpp]: Yeah lets get rid of it. We can always add it back if we want to. Also, the cc done printf isn't necessary either.

File: include/graph.h [hunk 6]
```diff
@@ -1,165 +1,162 @@
...
+class Graph {
+  node_id_t num_nodes;
+  long seed;
+  bool update_locked = false;
+  bool modified = false;
+  // a set containing one "representative" from each supernode
+  std::set<node_id_t>* representatives;
+  Supernode** supernodes;
+  // DSU representation of supernode relationship
+  node_id_t * parent;
```
Comment (etwest) [inline include/graph.h]: Style nit: Remove space after *

File: include/graph.h (hunk 1 above)
Comment (etwest) [reply include/graph.h:157]: Can you put in a pull request for the experiment repo that changes the query_expr in main to use these time hooks?

File: src/graph.cpp [hunk 7]
```diff
@@ -186,25 +186,25 @@ inline std::vector<std::vector<node_id_t>> Graph::supernodes_to_merge(std::pair<
...
 inline void Graph::merge_supernodes(Supernode** copy_supernodes, std::vector<node_id_t> &new_reps,
-               std::vector<std::vector<node_id_t>> &to_merge, bool make_copy, bool first_round) {
+               std::vector<std::vector<node_id_t>> &to_merge, bool first_round, bool make_copy) {
   bool except = false;
   std::exception_ptr err;
   // loop over the to_merge vector and perform supernode merging
   #pragma omp parallel for default(shared)
-  for (node_id_t i = 0; i < new_reps.size(); i++) {
-    node_id_t a = new_reps[i]; // fix for ubuntu18.04 ('for each' style in omp no good)
+  for (node_id_t i = 0; i < new_reps.size(); i++) { // NOLINT(modernize-loop-convert)
```
Comment (etwest) [inline src/graph.cpp:206]: What exactly does this comment mean?

File: src/graph.cpp [hunk 8]
```diff
@@ -186,25 +186,25 @@ inline std::vector<std::vector<node_id_t>> Graph::supernodes_to_merge(std::pair<
...
-    if (to_merge[a].size() != 0) new_reps.push_back(a);
+    if (!to_merge[a].empty()) new_reps.push_back(a);
 
   reps = new_reps;
   return to_merge;
 }
 
 inline void Graph::merge_supernodes(Supernode** copy_supernodes, std::vector<node_id_t> &new_reps,
-               std::vector<std::vector<node_id_t>> &to_merge, bool make_copy, bool first_round) {
+               std::vector<std::vector<node_id_t>> &to_merge, bool first_round, bool make_copy) {
```
Comment (etwest) [inline src/graph.cpp]: Nice catch. That might have explained part of the difference in performance we were observing.

File: src/graph.cpp [hunk 9]
```diff
@@ -242,21 +243,42 @@ std::vector<std::set<node_id_t>> Graph::boruvka_emulation(bool make_copy) {
...
-    modified = false;
-    sample_supernodes(query, reps);
-    std::vector<std::vector<node_id_t>> to_merge = supernodes_to_merge(query, reps);
-
-    // make a copy if necessary
-    if(make_copy && first_round) {
-      backed_up = reps;
-      if (!copy_in_mem) backup_to_disk(backed_up);
+  // function to restore supernodes after CC if make_copy is specified
+  auto cleanup_copy = [&make_copy, this, &backed_up, &copy_supernodes]() {
```
Comment (etwest) [inline src/graph.cpp:248]: This code looks right. I'm curious if you've tried inducing an exception to ensure the super-nodes are restored correctly. Maybe this should be a unit test?

File: src/graph.cpp (hunk 8 above)
Comment (etwest) [reply src/graph.cpp]: Actually, semantically it probably doesn't make a difference. I think we can (and should) reduce the number of arguments to this function to just one: namely 'make_copy' and pass in whether we require the supernodes to be backed up. (first_round && make_copy) or just first_round which starts true only if make_copy is true.

File: src/graph.cpp [hunk 10]
```diff
@@ -313,17 +323,26 @@ void Graph::restore_from_disk(std::vector<node_id_t> ids_to_restore) {
...
-  flush_call = std::chrono::steady_clock::now();
+  flush_start = std::chrono::steady_clock::now();
   bf->force_flush(); // flush everything in buffering system to make final updates
   GraphWorker::pause_workers(); // wait for the workers to finish applying the updates
-  flush_return = std::chrono::steady_clock::now();
+  flush_end = std::chrono::steady_clock::now();
   // after this point all updates have been processed from the buffer tree
 
-  if (!cont)
+  if (!cont) {
```
Comment (etwest) [inline src/graph.cpp]: style nit: {} unnecessary

File: src/graph.cpp [hunk 11]
```diff
@@ -348,10 +370,10 @@ void Graph::write_binary(const std::string& filename) {
   // after this point all updates have been processed from the buffering system
 
   auto binary_out = std::fstream(filename, std::ios::out | std::ios::binary);
-  int fail_factor = Sketch::get_failure_factor();
-  binary_out.write((char*)&seed, sizeof(long));
-  binary_out.write((char*)&num_nodes, sizeof(uint64_t));
-  binary_out.write((char*)&fail_factor, sizeof(int));
+  auto fail_factor = Sketch::get_failure_factor();
```
Comment (etwest) [inline src/graph.cpp:373]: This makes me nervous. I suppose if it works then it works though.

File: include/graph.h (hunk 1 above)
Comment (Victor-C-Zhang) [reply include/graph.h:157]: Yeah, it’s in draft rn

File: src/graph.cpp (hunk 7 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp:206]: It’s a directive for CLion so it doesn’t mark this as a loop that should be modernized

File: src/graph.cpp (hunk 9 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp:248]: Yes, you’re right. To do so we’ll have to add a dependency that allows us to mock. I can work on that if you think it’s important.

File: src/graph.cpp (hunk 11 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp:373]: what part of it?

File: src/graph.cpp (hunk 7 above)
Comment (etwest) [reply src/graph.cpp:206]: Could you add the comment below back in. I want to explain why we aren't using the modern style.

File: src/graph.cpp (hunk 11 above)
Comment (etwest) [reply src/graph.cpp:373]: Using auto for a variable we're going about to write to disk. I guess I'm nervous the compiler might chose some weird larger or smaller type to put fail_factor in. Though I don't know if that's actually possible.

File: src/graph.cpp (hunk 9 above)
Comment (etwest) [reply src/graph.cpp:248]: I'll add it to the Trello and either you or someone else can be assigned to it. Doesn't have to be a part of this PR.

File: src/graph.cpp (hunk 11 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp:373]: auto takes the type of the RHS at compile time. it’s like if we looked up the type of get_failure_factor (uint32) and spliced it in

File: src/graph.cpp (hunk 8 above)
Comment (Victor-C-Zhang) [reply src/graph.cpp]: sure, ill make that change

File: src/graph.cpp (hunk 11 above)
Comment (etwest) [reply src/graph.cpp:373]: Gotcha, sounds good

**Review by etwest — APPROVED** [review]
LGTM

**Pull request merged**